Enforce extra validations when a competition is visible - Fixes #604

pingskills opened the pull request:
The default factory for a competition used to set showAtAll true without a delegate.
The extra validations make this default competition invalid. To change this the default
is now showAtAll false which matches what happens in the front end. A lot of tests needed
to be updated that created competitions because of this change.

Let me know if you think this is the right approach to adding validations when making a competition visible.

jonatanklosko commented on WcaOnRails/spec/models/competition_spec.rb:
```diff
@@ -311,33 +311,41 @@
+      [:cityName, :countryId, :venue, :venueAddress, :website, :latitude, :longitude].each do |field|
+        it "requires #{field} when setting #{action}" do
+          competition_with_delegate.public_send "#{field}=", ""
+          competition_with_delegate.public_send "#{action}=", true
```
Such an assignment is a bit cumbersome. Moreover there's a handy way to write these two lines at once:
```ruby
competition_with_delegate.assign_attributes field => "", action => true
```

jonatanklosko commented on WcaOnRails/spec/models/competition_spec.rb:
```diff
@@ -311,33 +311,41 @@
-    end
+      it "must have at least one event when setting #{action}" do
+        competition_with_delegate.eventSpecs = ""
+        competition_with_delegate.public_send "#{action}=", true
```
Same as above
```ruby
competition_with_delegate.assign_attributes eventSpecs: "", action => true
```

jonatanklosko commented on WcaOnRails/spec/models/competition_spec.rb:
```diff
@@ -311,33 +311,41 @@
+      it "requires both dates when setting #{action}" do
+        competition_with_delegate.start_date = ""
+        competition_with_delegate.end_date = ""
+        competition_with_delegate.public_send "#{action}=", true
```
Same as above, three lines at once! =)
```ruby
competition_with_delegate.assign_attributes start_date: "", end_date: "", action => true
```

jonatanklosko commented:
I'm wondering whether it make sense to let the competition by invisible and unconfirmed by default when using the factory. I saw a lot of calls to `:confirmed_and_visible` and I worry that it might become a bit cumbersome, because most of the tests that will be added in the future would probably need to be ad least visible. Maybe we should have two factories here: one for a *fully ready* competition, and one for a *new created* one. I'm not sure and would be glad to see other opinions. We just need to make it to be handy and concise in the future usage =)

pingskills commented:
Thanks @jonatanklosko, I've updated the tests with your suggestions.

jonatanklosko commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -41,22 +41,27 @@ class Competition < ActiveRecord::Base
     end
   end
 
+  def confirmedOrVisible?
```
Please use snake case here =) These `isConfirmed` and `showAtAll` are camelCased because the tables was created and used originally in PHP code.

jonatanklosko commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -394,7 +399,7 @@ def end_date
 
   private def dates_must_be_valid
-    if !isConfirmed && self.year == 0 && self.month == 0 && self.day == 0 && @endYear == 0 && self.endMonth == 0 && self.endDay == 0
+    if !isConfirmed && !showAtAll && self.year == 0 && self.month == 0 && self.day == 0 && @endYear == 0 && self.endMonth == 0 && self.endDay == 0
```
Since you added a new method let's make use of it
`!isConfirmed && !showAtAll` -> '!confirmed_or_visible?`
Furthermore I think we should refacor the rest too, like so:
```ruby
if !confirmed_or_visible? && [year, month, day, @endYear, endMonth, endDay].all? { |n| n == 0 }
```

jonatanklosko commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -41,22 +41,27 @@ class Competition < ActiveRecord::Base
   [:cityName, :countryId, :venue, :venueAddress, :website, :latitude, :longitude].each do |field|
-    validates field, presence: true, if: :isConfirmed?
+    validates field, presence: true, if: :confirmed_or_visible?
   end
```
Maybe some of things I'm commenting aren't related to your PR, but since they show up here I feel like changing them =) So here I would see
```ruby
validates :cityName, :countryId, :venue, :venueAddress, :website, :latitude, :longitude, presence: true, if: :confirmed_or_visible?
```

jonatanklosko commented on WcaOnRails/spec/controllers/notifications_controller_spec.rb:
```diff
@@ -73,7 +73,7 @@
       let!(:unconfirmed_competition) { FactoryGirl.create :competition, isConfirmed: false, showAtAll: false }
```
I guess that we no longer need `isConfirmed: false, showAtAll: false`, right?

jonatanklosko commented on WcaOnRails/spec/controllers/notifications_controller_spec.rb:
```diff
@@ -73,7 +73,7 @@
       let!(:unconfirmed_competition) { FactoryGirl.create :competition, isConfirmed: false, showAtAll: false }
       let!(:confirmed_competition) { FactoryGirl.create(:competition, :confirmed, showAtAll: false) }
```
`showAtAll: false` - same as above

jonatanklosko commented on WcaOnRails/spec/controllers/notifications_controller_spec.rb:
```diff
@@ -73,7 +73,7 @@
       let!(:unconfirmed_competition) { FactoryGirl.create :competition, isConfirmed: false, showAtAll: false }
       let!(:confirmed_competition) { FactoryGirl.create(:competition, :confirmed, showAtAll: false) }
       let!(:visible_confirmed_competition) { FactoryGirl.create(:competition, :confirmed, showAtAll: true) }
```
`:confirmed, showAtAll: true` -> `:confirmed_and_visible`

jonatanklosko commented:
Actually I came up with a nice idea. Since we already have this `:confirmed` trait, maybe we should keep it in parallel and use `:visible` trait too instead of the whole `:confirmed_and_visible`? It's surely better than using `showAtAll: true` =)

jonatanklosko commented:
Okay, that's all from me. The last post is just a proposition and I'm waiting for some feedback from others, because we all are going to use this frequently in the future =)

larspetrus commented on WcaOnRails/spec/factories/competitions.rb:
```diff
@@ -44,5 +44,10 @@
       isConfirmed true
     end
+
+    trait :confirmed_and_visible do
```
This is a little weird. There already is a `confirmed` trait, so a separate `visible` trait would make more sense to me. If you want both, you can pass in `[:confirmed, :visible]` to `FactoryGirl.create`, which is the same number of characters.

pingskills replied:
I thought about creating a :visible trait which sets `showAtAll true` but if it the call to FactoryGirl only uses this trait the competition will be invalid because it needs a delegate. I could add `with_delegate` to the visible trait but then if you call `[:confirmed, :visible]` do you get 2 delegates? Or is it fine to just set showAtAll for  the visible trait and then if you want a valid competition you need to add either `with_delelgate` or `confirmed`?

larspetrus replied:
I think/hope FactoryGirl manages things so each trait is only called once.

If not, this might not be worth changing.

jfly replied:
I can confirm that FactoryGirl seems to do the right thing here. With the following changes:

```
~/gitting/worldcubeassociation.org/WcaOnRails @kaladin> git diff
diff --git a/WcaOnRails/spec/factories/competitions.rb b/WcaOnRails/spec/factories/competitions.rb
index 3417731..c8579a6 100644
--- a/WcaOnRails/spec/factories/competitions.rb
+++ b/WcaOnRails/spec/factories/competitions.rb
@@ -18,7 +18,7 @@ FactoryGirl.define do
     venue "My backyard"
     venueAddress "My backyard street"
     website "https://www.worldcubeassociation.org"
-    showAtAll true
+    #<<<showAtAll true
 
     guests_enabled true
 
@@ -44,5 +44,10 @@ FactoryGirl.define do
       with_delegate
       isConfirmed true
     end
+
+    trait :visible do
+      with_delegate
+      showAtAll true
+    end
   end
 end
~/gitting/worldcubeassociation.org/WcaOnRails @kaladin> 
```

I see the following behavior:

```
irb(main):004:0> FactoryGirl.create(:competition, :confirmed).delegates.length
   (0.2ms)  BEGIN
  SQL (0.2ms)  INSERT INTO `Persons` (`id`, `name`, `countryId`, `gender`, `year`, `month`, `day`) VALUES ('2003FLEI03', 'Brittany Christiansen I', 'USA', 'm', 1966, 4, 4)
   (11.3ms)  COMMIT
   (0.1ms)  BEGIN
  Person Load (0.3ms)  SELECT  `Persons`.* FROM `Persons` WHERE `Persons`.`id` = '2003FLEI03' LIMIT 1
  User Exists (0.2ms)  SELECT  1 AS one FROM `users` WHERE `users`.`email` = BINARY 'ken_streich@farrell.biz' LIMIT 1
  User Load (0.2ms)  SELECT  `users`.* FROM `users` WHERE `users`.`wca_id` = '2003FLEI03' LIMIT 1
  User Load (0.2ms)  SELECT `users`.* FROM `users` WHERE `users`.`wca_id` = '2003FLEI03'
  SQL (0.2ms)  INSERT INTO `users` (`name`, `email`, `country_iso2`, `gender`, `dob`, `encrypted_password`, `wca_id`, `delegate_status`, `confirmed_at`, `created_at`, `updated_at`) VALUES ('Brittany Christiansen I', 'ken_streich@farrell.biz', 'US', 'm', '1966-04-04', '$2a$04$4ogO1bHMW2QlXaRzPiHEquk2JkT.NUkV1OIbji5kfHQZP7MlKrW9O', '2003FLEI03', 'delegate', '2016-05-17 03:02:31', '2016-05-17 03:02:31', '2016-05-17 03:02:31')
   (9.6ms)  COMMIT
   (0.1ms)  BEGIN
  Competition Exists (0.3ms)  SELECT  1 AS one FROM `Competitions` WHERE `Competitions`.`id` = BINARY 'FooComp32015' LIMIT 1
  SQL (0.3ms)  INSERT INTO `Competitions` (`name`, `cityName`, `countryId`, `information`, `eventSpecs`, `venue`, `venueAddress`, `website`, `registration_open`, `registration_close`, `isConfirmed`, `year`, `month`, `day`, `endMonth`, `endDay`, `cellName`, `id`) VALUES ('Foo Comp 3 2015', 'San Francisco', 'USA', 'Information!', '333 333oh', 'My backyard', 'My backyard street', 'https://www.worldcubeassociation.org', '2016-05-03 03:01:23', '2016-05-10 03:01:23', 1, 2015, 5, 17, 5, 17, 'Foo Comp 3 2015', 'FooComp32015')
  SQL (0.2ms)  INSERT INTO `competition_delegates` (`competition_id`, `delegate_id`, `created_at`, `updated_at`) VALUES ('FooComp32015', 3, '2016-05-17 03:02:31', '2016-05-17 03:02:31')
  User Load (0.3ms)  SELECT `users`.* FROM `users` INNER JOIN `competition_organizers` ON `users`.`id` = `competition_organizers`.`organizer_id` WHERE `competition_organizers`.`competition_id` = 'FooComp32015'
  SQL (0.2ms)  DELETE FROM `competition_organizers` WHERE `competition_organizers`.`competition_id` = 'FooComp32015' AND (1=1)
  SQL (0.2ms)  DELETE FROM `competition_delegates` WHERE `competition_delegates`.`competition_id` = 'FooComp32015' AND (`competition_delegates`.`delegate_id` != 3)
  SQL (0.1ms)  UPDATE `Results` SET `Results`.`competitionId` = 'FooComp32015' WHERE `Results`.`competitionId` = ''
  SQL (0.1ms)  UPDATE `Preregs` SET `Preregs`.`competitionId` = 'FooComp32015' WHERE `Preregs`.`competitionId` = ''
  SQL (0.1ms)  UPDATE `Scrambles` SET `Scrambles`.`competitionId` = 'FooComp32015' WHERE `Scrambles`.`competitionId` = ''
  SQL (0.1ms)  UPDATE `CompetitionsMedia` SET `CompetitionsMedia`.`competitionId` = 'FooComp32015' WHERE `CompetitionsMedia`.`competitionId` = ''
  SQL (0.1ms)  UPDATE `competition_delegates` SET `competition_delegates`.`competition_id` = 'FooComp32015' WHERE `competition_delegates`.`competition_id` = ''
  SQL (0.1ms)  UPDATE `competition_organizers` SET `competition_organizers`.`competition_id` = 'FooComp32015' WHERE `competition_organizers`.`competition_id` = ''
   (16.0ms)  COMMIT
=> 1
irb(main):005:0> FactoryGirl.create(:competition, :visible).delegates.length
   (0.1ms)  BEGIN
  SQL (0.2ms)  INSERT INTO `Persons` (`id`, `name`, `countryId`, `gender`, `year`, `month`, `day`) VALUES ('2003FLEI04', 'Tabitha Schultz DDS', 'USA', 'm', 1966, 4, 4)
   (11.2ms)  COMMIT
   (0.1ms)  BEGIN
  Person Load (0.3ms)  SELECT  `Persons`.* FROM `Persons` WHERE `Persons`.`id` = '2003FLEI04' LIMIT 1
  User Exists (0.2ms)  SELECT  1 AS one FROM `users` WHERE `users`.`email` = BINARY 'narciso_bednar@gleasonkozey.org' LIMIT 1
  User Load (0.2ms)  SELECT  `users`.* FROM `users` WHERE `users`.`wca_id` = '2003FLEI04' LIMIT 1
  User Load (0.2ms)  SELECT `users`.* FROM `users` WHERE `users`.`wca_id` = '2003FLEI04'
  SQL (0.2ms)  INSERT INTO `users` (`name`, `email`, `country_iso2`, `gender`, `dob`, `encrypted_password`, `wca_id`, `delegate_status`, `confirmed_at`, `created_at`, `updated_at`) VALUES ('Tabitha Schultz DDS', 'narciso_bednar@gleasonkozey.org', 'US', 'm', '1966-04-04', '$2a$04$lOPKlmhwZfppTx1ZogBcaOH/AABvRtGaxTiYvdhabT6YlTWjXuuxy', '2003FLEI04', 'delegate', '2016-05-17 03:02:36', '2016-05-17 03:02:36', '2016-05-17 03:02:36')
   (4.0ms)  COMMIT
   (0.1ms)  BEGIN
  Competition Exists (0.3ms)  SELECT  1 AS one FROM `Competitions` WHERE `Competitions`.`id` = BINARY 'FooComp42015' LIMIT 1
  SQL (0.3ms)  INSERT INTO `Competitions` (`name`, `cityName`, `countryId`, `information`, `eventSpecs`, `venue`, `venueAddress`, `website`, `registration_open`, `registration_close`, `showAtAll`, `year`, `month`, `day`, `endMonth`, `endDay`, `cellName`, `id`) VALUES ('Foo Comp 4 2015', 'San Francisco', 'USA', 'Information!', '333 333oh', 'My backyard', 'My backyard street', 'https://www.worldcubeassociation.org', '2016-05-03 03:01:23', '2016-05-10 03:01:23', 1, 2015, 5, 17, 5, 17, 'Foo Comp 4 2015', 'FooComp42015')
  SQL (0.2ms)  INSERT INTO `competition_delegates` (`competition_id`, `delegate_id`, `created_at`, `updated_at`) VALUES ('FooComp42015', 4, '2016-05-17 03:02:36', '2016-05-17 03:02:36')
  User Load (0.4ms)  SELECT `users`.* FROM `users` INNER JOIN `competition_organizers` ON `users`.`id` = `competition_organizers`.`organizer_id` WHERE `competition_organizers`.`competition_id` = 'FooComp42015'
  SQL (0.2ms)  DELETE FROM `competition_organizers` WHERE `competition_organizers`.`competition_id` = 'FooComp42015' AND (1=1)
  SQL (0.2ms)  DELETE FROM `competition_delegates` WHERE `competition_delegates`.`competition_id` = 'FooComp42015' AND (`competition_delegates`.`delegate_id` != 4)
  SQL (0.1ms)  UPDATE `Results` SET `Results`.`competitionId` = 'FooComp42015' WHERE `Results`.`competitionId` = ''
  SQL (0.1ms)  UPDATE `Preregs` SET `Preregs`.`competitionId` = 'FooComp42015' WHERE `Preregs`.`competitionId` = ''
  SQL (0.1ms)  UPDATE `Scrambles` SET `Scrambles`.`competitionId` = 'FooComp42015' WHERE `Scrambles`.`competitionId` = ''
  SQL (0.1ms)  UPDATE `CompetitionsMedia` SET `CompetitionsMedia`.`competitionId` = 'FooComp42015' WHERE `CompetitionsMedia`.`competitionId` = ''
  SQL (0.1ms)  UPDATE `competition_delegates` SET `competition_delegates`.`competition_id` = 'FooComp42015' WHERE `competition_delegates`.`competition_id` = ''
  SQL (0.2ms)  UPDATE `competition_organizers` SET `competition_organizers`.`competition_id` = 'FooComp42015' WHERE `competition_organizers`.`competition_id` = ''
   (3.7ms)  COMMIT
=> 1
irb(main):006:0> FactoryGirl.create(:competition, :confirmed, :visible).delegates.length
   (0.2ms)  BEGIN
  SQL (0.2ms)  INSERT INTO `Persons` (`id`, `name`, `countryId`, `gender`, `year`, `month`, `day`) VALUES ('2003FLEI05', 'Adeline Zemlak', 'USA', 'm', 1966, 4, 4)
   (11.7ms)  COMMIT
   (0.1ms)  BEGIN
  Person Load (0.3ms)  SELECT  `Persons`.* FROM `Persons` WHERE `Persons`.`id` = '2003FLEI05' LIMIT 1
  User Exists (0.2ms)  SELECT  1 AS one FROM `users` WHERE `users`.`email` = BINARY 'phoebe_dare@satterfield.com' LIMIT 1
  User Load (0.4ms)  SELECT  `users`.* FROM `users` WHERE `users`.`wca_id` = '2003FLEI05' LIMIT 1
  User Load (0.2ms)  SELECT `users`.* FROM `users` WHERE `users`.`wca_id` = '2003FLEI05'
  SQL (0.2ms)  INSERT INTO `users` (`name`, `email`, `country_iso2`, `gender`, `dob`, `encrypted_password`, `wca_id`, `delegate_status`, `confirmed_at`, `created_at`, `updated_at`) VALUES ('Adeline Zemlak', 'phoebe_dare@satterfield.com', 'US', 'm', '1966-04-04', '$2a$04$u8FUJEDfxCOH5KOTPi7Bc.cqweaAae7oPErmYFonw5OfF.Tj1LsZi', '2003FLEI05', 'delegate', '2016-05-17 03:02:41', '2016-05-17 03:02:41', '2016-05-17 03:02:41')
   (9.6ms)  COMMIT
   (0.2ms)  BEGIN
  Competition Exists (0.2ms)  SELECT  1 AS one FROM `Competitions` WHERE `Competitions`.`id` = BINARY 'FooComp52015' LIMIT 1
  SQL (0.2ms)  INSERT INTO `Competitions` (`name`, `cityName`, `countryId`, `information`, `eventSpecs`, `venue`, `venueAddress`, `website`, `registration_open`, `registration_close`, `isConfirmed`, `showAtAll`, `year`, `month`, `day`, `endMonth`, `endDay`, `cellName`, `id`) VALUES ('Foo Comp 5 2015', 'San Francisco', 'USA', 'Information!', '333 333oh', 'My backyard', 'My backyard street', 'https://www.worldcubeassociation.org', '2016-05-03 03:01:23', '2016-05-10 03:01:23', 1, 1, 2015, 5, 17, 5, 17, 'Foo Comp 5 2015', 'FooComp52015')
  SQL (0.1ms)  INSERT INTO `competition_delegates` (`competition_id`, `delegate_id`, `created_at`, `updated_at`) VALUES ('FooComp52015', 5, '2016-05-17 03:02:41', '2016-05-17 03:02:41')
  User Load (0.2ms)  SELECT `users`.* FROM `users` INNER JOIN `competition_organizers` ON `users`.`id` = `competition_organizers`.`organizer_id` WHERE `competition_organizers`.`competition_id` = 'FooComp52015'
  SQL (0.1ms)  DELETE FROM `competition_organizers` WHERE `competition_organizers`.`competition_id` = 'FooComp52015' AND (1=1)
  SQL (0.2ms)  DELETE FROM `competition_delegates` WHERE `competition_delegates`.`competition_id` = 'FooComp52015' AND (`competition_delegates`.`delegate_id` != 5)
  SQL (0.1ms)  UPDATE `Results` SET `Results`.`competitionId` = 'FooComp52015' WHERE `Results`.`competitionId` = ''
  SQL (0.1ms)  UPDATE `Preregs` SET `Preregs`.`competitionId` = 'FooComp52015' WHERE `Preregs`.`competitionId` = ''
  SQL (0.1ms)  UPDATE `Scrambles` SET `Scrambles`.`competitionId` = 'FooComp52015' WHERE `Scrambles`.`competitionId` = ''
  SQL (0.1ms)  UPDATE `CompetitionsMedia` SET `CompetitionsMedia`.`competitionId` = 'FooComp52015' WHERE `CompetitionsMedia`.`competitionId` = ''
  SQL (0.1ms)  UPDATE `competition_delegates` SET `competition_delegates`.`competition_id` = 'FooComp52015' WHERE `competition_delegates`.`competition_id` = ''
  SQL (0.1ms)  UPDATE `competition_organizers` SET `competition_organizers`.`competition_id` = 'FooComp52015' WHERE `competition_organizers`.`competition_id` = ''
   (15.9ms)  COMMIT
=> 1
```

Note how `FactoryGirl.create(:competition, :confirmed, :visible)` only created 1 delegate, not two.

jfly replied:
In light of all that, I support @larspetrus's idea to split `confirmed_and_visible` into two separate traits.

pingskills commented:
Do we ever want a competition visible but not confirmed? Maybe we should not have a separate attribute but rather a status that can be "unconfirmed", "confirmed", and "visible"?

jfly commented:
Yes, the board uses "not confirmed, but visible" to let delegates make changes to a competition after it is announced. I don't like this, and would rather that the delegate always be able to send in changes that the board has to approve, but that's a big engineering effort.

pedrosino commented:
Just for the record, we don't do that anymore. The delegates send an email with the requested change and we do it ourselves.

jfly commented:
@pingskills, in light of what @pedrosino said, your idea sounds like a good idea for potential future cleanup =)

pingskills commented:
As suggested I've updated the competition factory to have a `visible` trait which has `with_delegate` and `showAtAll true`. I've updated the tests to use the `:confirmed` and `:visible` traits where appropriate.

jfly commented:
LGTM!

jfly commented:
Thanks, Jeff!
